hazelcast/src/main/java/com/hazelcast/cluster/impl/ClusterStateManager.java

```diff
@@ -284,4 +285,38 @@ public String toString() {
         sb.append('}');
         return sb.toString();
     }
+
+    private static final class StateManagerExceptionHandler implements FutureUtil.ExceptionHandler {
+        private final ILogger logger;
+        private Throwable error;
```

Comment:
I would make this field volatile. Afaik it is set by one thread and the rethrowIfFailed is called by another. Perhaps it lifts on top of an existing happens before relation, but it isn't obvious. So I would place it in a volatile variable to prevent any confusion.